src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs

```diff
@@ -218,18 +217,19 @@ public void ExecuteInitializeMethod(HostSessionStartAnalysisScope sessionScope,
         /// <param name="cancellationToken">Cancellation token.</param>
         public void ExecuteCompilationStartActions(ImmutableArray<CompilationStartAnalyzerAction> actions, HostCompilationStartAnalysisScope compilationScope, CancellationToken cancellationToken)
         {
+            // This context doesn't build up any state as we pass it to the Action method of the analyzer. As such, we
+            // can use the same instance across all actions.
+            var context = new AnalyzerCompilationStartAnalysisContext(compilationScope,
+                Compilation, AnalyzerOptions, _compilationAnalysisValueProviderFactory, cancellationToken);
+            var contextInfo = new AnalysisContextInfo(Compilation);
+
             foreach (var startAction in actions)
             {
-                cancellationToken.ThrowIfCancellationRequested();
```

Comment:
cancellation check moved into ExecuteAndCatchIfThrows for consistency and simplicity for all callers.